### PR TITLE
Implementation of Cef.RegisterWidevineCdm

### PR DESCRIPTION
**Pull request by spazzarama**
Implementation of Cef.RegisterWidevineCdm with test as per feature request #1934

I have completed testing, however in order for others to test you need to:
 1. have access to WidevineCdm binaries (it might be possible test using Chrome copy, possibly with manifest.json changes - see #1934 for list of files required and example manifest.json)
 1. uncomment the call to Cef.RegisterWidevineCdm(@".\WidevineCdm"); in CefExample.Init
 1. place the Widevine binaries in a folder called 'WidevineCdm' within the CefSharp.Wpf.Example build directory (e.g. bin\x64\Debug\WidevineCdm)

Please advise if any changes needed, styling issues, etc...

Description of changes:
 * Added types CefSharp/CdmRegistration.cs (return value of registration process) and enumeration  CefSharp/CdmRegistrationErrorCode.cs to CefSharp
 * Added RegisterWidevineCdm metho to Cef.h in CefSharp.Core
 * Added CefRegisterCdmCallbackWrapper.h to CefSharp.Core project

### CODE REVIEW

File: CefSharp.Core/Cef.h [hunk 1]
```diff
@@ -607,5 +608,60 @@ namespace CefSharp
...
+        ///
+        /// |callback| will be executed asynchronously once registration is complete.
+        ///
+        /// On Linux this function must be called before CefInitialize() and the
+        /// registration cannot be changed during runtime. If registration is not
+        /// supported at the time that CefRegisterWidevineCdm() is called then |callback|
+        /// will receive a |result| value of CEF_CDM_REGISTRATION_ERROR_NOT_SUPPORTED.
+        /// </summary>
+        /// <returns>Returns the CDM registration result.</returns>
+        static Task<CdmRegistration^>^ RegisterWidevineCdm(String^ path)
```
Comment (amaitland) [inline CefSharp.Core/Cef.h]: I know a similar pattern has been used in the past, I've revised my take on this, would prefer to wrap the underlying `API` as close to `1:1` as possible and implement any async wrappers in `C#`. 

Please add `IRegisterCdmCallback` as a param and return `void` here.

File: CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#include "CefWrapper.h"
+
+namespace CefSharp
+{
+    namespace Internals
+    {
+        private class CefRegisterCdmCallbackWrapper : public CefRegisterCdmCallback
+        {
+        private:
+            gcroot<TaskCompletionSource<CdmRegistration^>^> _taskCompletionSource;
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h]: Replace with `IRegisterCdmCallback`

File: CefSharp.Example/Resources/CdmSupportTest.html [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        function doTest() {
+            probeSupport().then(function (support) {
+                printSupport(JSON.stringify(support, null, '  '));
+            });
+        }
+    </script>
+</head>
+<body onload="doTest()">
```
Comment (amaitland) [inline CefSharp.Example/Resources/CdmSupportTest.html]: Should probably be some sort of disclaimer here that explains `widevine CDM`  requires additional steps to enable and `propritary` codes are not supported in the default `Nuget` packages as they require licensing.

What happens if you open this page with a default build?

File: CefSharp.Example/CefExample.cs [hunk 4]
```diff
@@ -41,6 +42,10 @@ public static void Init(bool osr, bool multiThreadedMessageLoop, IBrowserProcess
             // Environment.SetEnvironmentVariable("GOOGLE_DEFAULT_CLIENT_ID", "");
             // Environment.SetEnvironmentVariable("GOOGLE_DEFAULT_CLIENT_SECRET", "");
 
+            // Widevine CDM registration - pass in directory where Widevine CDM binaries and manifest.json are located.
+            // To organize access to Widevine binaries: https://www.widevine.com/contact.html
```
Comment (amaitland) [inline CefSharp.Example/CefExample.cs]: Should link to the issue #1934

**Comment by spazzarama** [comment]
I have just noticed that I didn't postfix the RegisterWidevineCdm method name with "Async".

Please advise if this should be the case (it returns a `Task<CdmRegistration^>^`).

File: CefSharp.Core/Cef.h (hunk 1 above)
Comment (amaitland) [inline CefSharp.Core/Cef.h]: By all means add an overload that has the same signature

Comment (amaitland) [inline CefSharp.Core/Cef.h]: Overload should be `RegisterWidevineCdmAsync`, you are correct.

File: CefSharp.Example/Resources/CdmSupportTest.html (hunk 3 above)
Comment (spazzarama) [inline CefSharp.Example/Resources/CdmSupportTest.html]: If Widevine not registered and not using proprietary codecs just says that widevine is "NOT AVAILABLE", and some of the codecs are listed as not supporting playback.

I'll add a note at the end of the page re proprietary codecs and contacting Widevine for access to widevine.

File: CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h (hunk 2 above)
Comment (spazzarama) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h]: Do you want a base concrete implementation of IRegisterCdmCallback as well or just implement an example within the example project?

Comment (amaitland) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h]: If you wish to have `Cef.RegisterWidevineCdmAsync` then add a concrete implementation to the `CefSharp` project.

Comment (spazzarama) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h]: Sorry to be a pain, I'm fairly new to the CefSharp codebase and not that familiar with other areas of it.

In order to call the underlying CefRegisterWidevineCdm it wants a CefRegisterCdmCallback (defined in the Cef codebase). Can you please point me to an example where a similar process takes place but using the interface as you described?

Or are you saying pass our own defined interface into the callback wrapper and use that to trigger back through to the caller? But still use the CefRegisterCdmCallback internally.

Again, sorry for the noobness!

Comment (spazzarama) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h]: Never mind I get it, just replace the _taskCompletionSource with it. :)

Comment (amaitland) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackWrapper.h]: > Never mind I get it, just replace the _taskCompletionSource with it. :)

Pretty much, see https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/Internals/CefCompletionCallbackAdapter.h for an example.

Also `IRegisterCdmCallback` should implement `IDisposable`

**Comment by spazzarama** [comment]
@amaitland your review comments should now all be addressed:

 * added IRegisterCdmCallback and simple concrete class for convenience.
 * Cef.RegisterWidevineCdm async result replaced with IRegisterCdmCallback (one override with no parameter still provided as often the result of the process will be of no interest - this also keeps the example simple)
 * CdmSupportTest.html includes links to additional setup steps for both Widevine and proprietary codecs
 * CefExample.cs links to issue for more details. The call to Cef.RegisterWidevineCdm is left uncommented as it has no impact if the directory does not exist, this means someone could try out the Wpf Example without recompiling and just placing the widevine package in place.

File: CefSharp.Core/Cef.h [hunk 5]
```diff
@@ -607,5 +608,67 @@ namespace CefSharp
...
+        /// |callback| will be executed asynchronously once registration is complete.
+        ///
+        /// On Linux this function must be called before CefInitialize() and the
+        /// registration cannot be changed during runtime. If registration is not
+        /// supported at the time that CefRegisterWidevineCdm() is called then |callback|
+        /// will receive a |result| value of CEF_CDM_REGISTRATION_ERROR_NOT_SUPPORTED.
+        /// </summary>
+        static void RegisterWidevineCdm(String^ path, IRegisterCdmCallback^ callback)
+        {
+            auto adapter = new CefRegisterCdmCallbackAdapter(callback);
```
Comment (amaitland) [inline CefSharp.Core/Cef.h]: There's no point creating a `CefRegisterCdmCallbackAdapter` instance of `callback` is null, so add `nullptr` check here and remove the `RegisterWidevineCdm` overload.

File: CefSharp.Core/Internals/CefRegisterCdmCallbackAdapter.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+        public:
+            CefRegisterCdmCallbackAdapter(IRegisterCdmCallback^ callback)
+            {
+                _callback = callback;
+            }
+
+            ~CefRegisterCdmCallbackAdapter()
+            {
+                if (static_cast<IRegisterCdmCallback^>(_callback) != nullptr)
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackAdapter.h]: You can call `delete` on a `nullptr`, ends up being a `noop`, so no need to check this.

File: CefSharp.Core/Internals/CefRegisterCdmCallbackAdapter.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            /// </summary>
+            virtual void OnCdmRegistrationComplete(cef_cdm_registration_error_t result,
+                const CefString& error_message) OVERRIDE
+            {
+                auto r = gcnew CdmRegistration();
+
+                r->ErrorCode = (CdmRegistrationErrorCode)result;
+                r->ErrorMessage = StringUtils::ToClr(error_message);
+
+                if (static_cast<IRegisterCdmCallback^>(_callback) != nullptr)
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefRegisterCdmCallbackAdapter.h]: Better to only create an instance if there is a callback and remove this check.

File: CefSharp.Core/Cef.h [hunk 8]
```diff
@@ -607,5 +608,67 @@ namespace CefSharp
...
+        ///   2. widevinecdm file from the CDM binary distribution (e.g.
+        ///      widevinecdm.dll on Windows, libwidevinecdm.dylib on OS X,
+        ///      libwidevinecdm.so on Linux).
+        ///   3. widevidecdmadapter file from the CEF binary distribution (e.g.
+        ///      widevinecdmadapter.dll on Windows, widevinecdmadapter.plugin on OS X,
+        ///      libwidevinecdmadapter.so on Linux).
+        ///
+        /// If any of these files are missing or if the manifest file has incorrect
+        /// contents the registration will fail and |callback| will receive a |result|
+        /// value of CEF_CDM_REGISTRATION_ERROR_INCORRECT_CONTENTS.
```
Comment (amaitland) [inline CefSharp.Core/Cef.h]: Comments should be updated to make relevant in context of `CefSharp`.

File: CefSharp/RegisterCdmCallback.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+using System;
+
+namespace CefSharp
+{
+    /// <summary>
+    /// Provides a simple callback implementation of <see cref="IRegisterCdmCallback"/> for use with Widevine CDM registration.
+    /// </summary>
+    public class RegisterCdmCallback: IRegisterCdmCallback
```
Comment (amaitland) [inline CefSharp/RegisterCdmCallback.cs]: Having the default implementation implement a `TaskCompletionSource` wrapper would be consistent with the rest of the codebase and allow you to add `Cef.RegisterWidevineCdmAsync`

**Comment by amaitland** [comment]
@spazzarama Few more minor changes required 👍

File: CefSharp.Core/Cef.h (hunk 5 above)
Comment (spazzarama) [inline CefSharp.Core/Cef.h]: You mean pass nullptr into CefRegisterWidevineCdm when there is no callback provided? (tested and yep that doesn't cause any issues with the CEF API - at this time anyway).

Comment (amaitland) [inline CefSharp.Core/Cef.h]: > /*--cef(optional_param=callback)--*/

The `CEF API` says it's optional, so yes pass in `NULL` when there is no callback.

Comment (amaitland) [inline CefSharp.Core/Cef.h]: https://bitbucket.org/chromiumembedded/cef/commits/a1fc6f1#Linclude/cef_web_plugin.hT210

**Comment by spazzarama** [comment]
@amaitland those changes should be addressed now. Let me know if anything else needs to be sorted.

**Comment by amaitland** [comment]
@spazzarama Merged with the following changes https://github.com/cefsharp/CefSharp/commit/ee5fd34ee66933e6686ca7e354409821a018bc28

**Pull request merged**